clique/src/service.rs

```diff
@@ -205,12 +208,94 @@ where
+            AddConnection(public_key, address) => {
+                if self.manager.add_peer(public_key.clone(), address.clone()) {
+                    self.spawn_new_outgoing(public_key, address, result_for_parent.clone());
+                };
+            }
+            // remove the peer from the manager all workers will be killed automatically, due to closed channels
+            DelConnection(public_key) => {
+                self.manager.remove_peer(&public_key);
+            }
+            // pass the data to the manager
+            SendData(data, public_key) => match self.manager.send_to(&public_key, data) {
+                Ok(_) => trace!(target: LOG_TARGET, "Sending data to {}.", public_key),
+                Err(e) => trace!(
+                    target: LOG_TARGET,
+                    "Failed sending to {}: {}",
+                    public_key,
+                    e
+                ),
+            },
+        }
+    }
+
+    fn handle_authorization_request(
+        &mut self,
+        public_key: <SK as SecretKey>::PublicKey,
+        response_channel: oneshot::Sender<bool>,
+    ) {
+        let authorization_result = self.manager.is_authorized(&public_key);
+        if response_channel.send(authorization_result).is_err() {
+            warn!(
+                target: LOG_TARGET,
+                "Other side of the Authorization Service is already closed."
+            );
+        }
+    }
+
+    fn handle_data_for_network(
+        &mut self,
+        public_key: <SK as SecretKey>::PublicKey,
+        maybe_data_for_network: Option<UnboundedSender<D>>,
+        result_for_parent: &UnboundedSender<(
+            <SK as SecretKey>::PublicKey,
+            Option<UnboundedSender<D>>,
+        )>,
+    ) {
+        use AddResult::*;
+        match maybe_data_for_network {
+            Some(data_for_network) => {
+                match self.add_connection(public_key.clone(), data_for_network) {
+                    Uninterested => warn!(
+                        target: LOG_TARGET,
+                        "Established connection with peer {} for unknown reasons.", public_key
+                    ),
+                    Added => info!(
+                        target: LOG_TARGET,
+                        "New connection with peer {}.", public_key
+                    ),
+                    Replaced => info!(
+                        target: LOG_TARGET,
+                        "Replaced connection with peer {}.", public_key
+                    ),
+                }
+            }
+            None => {
+                if let Some(address) = self.peer_address(&public_key) {
+                    self.spawn_new_outgoing(public_key, address, result_for_parent.clone());
+                }
+            }
+        }
+    }
+
     /// Run the service until a signal from exit.
-    pub async fn run(mut self, mut exit: oneshot::Receiver<()>) {
+    pub async fn run(mut self, mut exit: oneshot::Receiver<()>) -> Result<(), String> {
```

Comment:
I have mixed feelings with this. Only thing we can do with these errors is logging, we do not handle them in any way. I can introduce a new error that will be just a wrapper for `String`. Then I will have to write `Into/From` for it for `&str` conversions for the `?` operator, etc. It will be pure boilerplate. For these errors we don't even care about backtraces. Perhaps just use `Error` from anyhow (new dependency)?

Reply:
I think defining the errors using `thiserror` could make sense. This way you have proper types implementing `Error` and in the future you can handle this errors easily. But let's wait for @timorleph

Reply:
Just normal enum errors are much nicer, lets not add another layer of problems in case we ever manage to get to implement actual tests for these things (in which case "stringy" errors would be extremely annoying). If you are worried about the boilerplate of defining all the `Display`s and whatnot then @DamianStraszak 's suggestion of `thiserror` is reasonable – tbh we probably should've been using it all along (at the very least in node, ABFT might have some argument about minimizing deps).

Reply:
Well, I guess we are not addint `thiserror` then, oh well, maybe next time or when we have some time for a refactor...